File: backend/pkg/core/page/service.go
package page

import (
	"context"
	"math"

	"backend/pkg/core"
	"backend/pkg/types"

	"github.com/google/uuid"
)

// Service provides business logic for pages
type Service struct {
	pageStore PageStore
}

// NewService creates a new page service with the provided store
func NewService(pageStore PageStore) *Service {
	return &Service{
		pageStore: pageStore,
	}
}

// GetByID retrieves a page by its ID
func (s *Service) GetByID(ctx context.Context, id uuid.UUID) (*types.Page, error) {
	return s.pageStore.FindByID(ctx, id)
}

// GetBySlug retrieves a page by its slug
func (s *Service) GetBySlug(ctx context.Context, slug string) (*types.Page, error) {
	return s.pageStore.FindBySlug(ctx, slug)
}

// List retrieves pages with pagination and optional filters
func (s *Service) List(ctx context.Context, pageNum, perPage int, isPublished *bool) (*ListResult, error) {
	// Apply defaults
	if perPage <= 0 {
		perPage = 20
	}
	if pageNum <= 0 {
		pageNum = 1
	}

	opts := types.PageListOptions{
		Page:        pageNum,
		PerPage:     perPage,
		IsPublished: isPublished,
	}

	pages, total, err := s.pageStore.List(ctx, opts)
	if err != nil {
		return nil, err
	}

	totalPages := int(math.Ceil(float64(total) / float64(perPage)))

	return &ListResult{
		Pages:      pages,
		Total:      total,
		Page:       pageNum,
		PerPage:    perPage,
		TotalPages: totalPages,
	}, nil
}

// Create creates a new page
func (s *Service) Create(ctx context.Context, req CreateRequest) (*types.Page, error) {
	// Check if slug already exists
	existing, err := s.pageStore.FindBySlug(ctx, req.Slug)
	if err != nil && err != core.ErrNotFound {
		return nil, err
	}
	if existing != nil {
		return nil, core.ErrAlreadyExists
	}

	page := &types.Page{
		ID:          uuid.New(),
		Slug:        req.Slug,
		Title:       req.Title,
		Content:     req.Content,
		Description: req.Description,
		ImageURL:    req.ImageURL,
		MetaData:    req.MetaData,
		IsPublished: req.IsPublished,
	}

	if err := s.pageStore.Save(ctx, page); err != nil {
		return nil, err
	}

	return page, nil
}

// Update updates an existing page
func (s *Service) Update(ctx context.Context, id uuid.UUID, req UpdateRequest) (*types.Page, error) {
	page, err := s.pageStore.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}

	// Apply updates
	if req.Title != nil {
		page.Title = *req.Title
	}
	if req.Content != nil {
		page.Content = *req.Content
	}
	if req.Description != nil {
		page.Description = *req.Description
	}
	if req.ImageURL != nil {
		page.ImageURL = *req.ImageURL
	}
	if req.MetaData != nil {
		page.MetaData = *req.MetaData
	}
	if req.IsPublished != nil {
		page.IsPublished = *req.IsPublished
	}

	if err := s.pageStore.Save(ctx, page); err != nil {
		return nil, err
	}

	return page, nil
}

// Delete removes a page by its ID
func (s *Service) Delete(ctx context.Context, id uuid.UUID) error {
	return s.pageStore.Delete(ctx, id)
}

File: backend/src/core/page/service.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::error::AppError;

use super::{
    Page, PageCreateRequest, PageListOptions, PageListResult, PageRepository, PageUpdateRequest,
};

#[derive(Clone)]
pub struct PageService {
    repository: Arc<dyn PageRepository>,
}

impl PageService {
    pub fn new(repository: Arc<dyn PageRepository>) -> Self {
        Self { repository }
    }

    pub async fn get_by_id(&self, id: Uuid) -> Result<Page, AppError> {
        self.repository.find_by_id(id).await
    }

    pub async fn get_by_slug(&self, slug: &str) -> Result<Page, AppError> {
        self.repository.find_by_slug(slug).await
    }

    pub async fn list(
        &self,
        page: i64,
        per_page: i64,
        is_published: Option<bool>,
    ) -> Result<PageListResult, AppError> {
        let page = if page <= 0 { 1 } else { page };
        let per_page = if per_page <= 0 { 20 } else { per_page };
        let (pages, total) = self
            .repository
            .list(PageListOptions {
                page,
                per_page,
                is_published,
            })
            .await?;
        let total_pages = if total == 0 {
            0
        } else {
            ((total - 1) / per_page) + 1
        };
        Ok(PageListResult {
            pages,
            total,
            page,
            per_page,
            total_pages,
        })
    }

    pub async fn create(&self, request: PageCreateRequest) -> Result<Page, AppError> {
        match self.repository.find_by_slug(&request.slug).await {
            Ok(_) => {
                return Err(AppError::Conflict("resource already exists".to_owned()));
            }
            Err(AppError::NotFound) => {}
            Err(error) => return Err(error),
        }

        let mut page = Page {
            id: Uuid::new_v4(),
            slug: request.slug,
            title: request.title,
            content: request.content,
            description: request.description,
            image_url: request.image_url,
            meta_data: request.meta_data,
            is_published: request.is_published,
            created_at: None,
            updated_at: None,
        };
        self.repository.save(&mut page).await?;
        Ok(page)
    }

    pub async fn update(&self, id: Uuid, request: PageUpdateRequest) -> Result<Page, AppError> {
        let mut page = self.repository.find_by_id(id).await?;
        if let Some(title) = request.title {
            page.title = title;
        }
        if let Some(content) = request.content {
            page.content = content;
        }
        if let Some(description) = request.description {
            page.description = description;
        }
        if let Some(image_url) = request.image_url {
            page.image_url = image_url;
        }
        if let Some(meta_data) = request.meta_data {
            page.meta_data = Some(meta_data);
        }
        if let Some(is_published) = request.is_published {
            page.is_published = is_published;
        }
        self.repository.save(&mut page).await?;
        Ok(page)
    }

    pub async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        self.repository.delete(id).await
    }
}
